File: src/utils/calendarEventsApi.js
import { apiFetch } from '../auth/AuthContext';

function apiError(res, data) {
    let msg = (data && data.error) || `HTTP ${res.status}`;
    if (res.status === 414) {
        msg = 'Request URI too long';
    }
    const err = new Error(msg);
    err.status = res.status;
    err.code = data && data.error;
    return err;
}

/** Fetch shared course events for the visible date range. */
export async function fetchEvents({ from, to, courses }) {
    const codes = [...new Set((courses || []).filter(Boolean))];
    const params = new URLSearchParams({ from, to });
    if (codes.length > 0) {
        params.set('courses', codes.join(','));
    }
    const res = await apiFetch(`/api/events?${params}`);
    const data = await res.json().catch(() => ({}));
    if (!res.ok) throw apiError(res, data);
    return Array.isArray(data.events) ? data.events : [];
}

export async function createEvent(partial) {
    const res = await apiFetch('/api/events', {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(partial),
    });
    const data = await res.json().catch(() => ({}));
    if (!res.ok) throw apiError(res, data);
    return data.event;
}

export async function patchEvent(id, partial) {
    const res = await apiFetch(`/api/events/${encodeURIComponent(id)}`, {
        method: 'PATCH',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(partial),
    });
    const data = await res.json().catch(() => ({}));
    if (!res.ok) throw apiError(res, data);
    return data.event;
}

export async function removeEvent(id) {
    const res = await apiFetch(`/api/events/${encodeURIComponent(id)}`, {
        method: 'DELETE',
    });
    const data = await res.json().catch(() => ({}));
    if (!res.ok) throw apiError(res, data);
    return data.id;
}

File: app/lib/api/calendar_events_api.dart
import '../models/calendar_event.dart';
import 'api_client.dart';

class CalendarEventsApi {
  CalendarEventsApi(this._client);
  final ApiClient _client;

  /// `GET /api/events?from=&to=&courses=` — shared course events.
  Future<List<CalendarEvent>> fetchEvents({
    required String from,
    required String to,
    required List<String> courses,
  }) async {
    final query = <String, dynamic>{'from': from, 'to': to};
    final unique = courses.where((c) => c.isNotEmpty).toSet().toList();
    if (unique.isNotEmpty) query['courses'] = unique.join(',');
    final data = await _client.requestJson('/api/events', query: query);
    final raw = data['events'];
    if (raw is List) {
      return raw
          .whereType<Map>()
          .map((e) => CalendarEvent.fromJson(Map<String, dynamic>.from(e)))
          .toList();
    }
    return const [];
  }

  Future<CalendarEvent> createEvent(Map<String, dynamic> payload) async {
    final data =
        await _client.requestJson('/api/events', method: 'POST', body: payload);
    return CalendarEvent.fromJson(Map<String, dynamic>.from(data['event']));
  }

  Future<CalendarEvent> patchEvent(String id, Map<String, dynamic> payload) async {
    final data = await _client.requestJson('/api/events/$id',
        method: 'PATCH', body: payload);
    return CalendarEvent.fromJson(Map<String, dynamic>.from(data['event']));
  }

  Future<void> removeEvent(String id) async {
    await _client.requestJson('/api/events/$id', method: 'DELETE');
  }
}
